### PR TITLE
[AIRFLOW-442]Added SFTPHook

### PR DESCRIPTION
**Pull request by biellls**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-442

Testing Done:
- Tested on a local SFTP server

### CODE REVIEW

**Comment by bolkedebruin** [comment]
Thanks! Please provide tests and documentation. Travis has a working ssh server, see sshhook tests.

**Comment by r39132** [comment]
Also, please fix style errors per landscape

**Comment by jonbender** [comment]
I'd recommend adding an option to use an SSH key pair, that is, if password is unspecified, allow users to provide a JSON blob in the Extra field eg.
`{"private_key": "/home/airflow/.ssh/id_rsa", "private_key_pass": "secret123"}` 
and pass those as args to the connection

File: airflow/contrib/hooks/sftp_hook.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+        return self.conn
+
+    def close_conn(self):
+        """
+        Closes the connection. An error will occur if the
+        connection wasnt ever opened.
+        """
+        conn = self.conn
+        conn.close()
```
Comment (cmkimerer) [inline airflow/contrib/hooks/sftp_hook.py:54]: Should you set `self.conn = None` after this?  I would have expected the following block of code to work:

```
hook = SFTPHook(conn_id)
hook.list_directory('/')
hook.close()
hook.get_conn() # note: this is implicit if you had just called the next line
hook.list_directory('/')
```

Comment (biellls) [reply airflow/contrib/hooks/sftp_hook.py:54]: As I mentioned before this hook is a direct translation from the ftp hook, so this function corresponds to https://github.com/OpenGov/airflow/blob/master/airflow/contrib/hooks/ftp_hook.py#L70. If this is not the right way to do it I will gladly change, but in this case ftp_hook should probably change too.

**Comment by biellls** [comment]
Thanks for the comments and sorry for the delay. I will try to fix style errors, add tests and implement @jonbender's suggestion ~~this weekend if~~ when I get the chance.

**Comment by criccomini** [comment]
We're doing a PR cleanup. Please respond within 5 days of your intentions regarding this PR and an ETA.

**Comment by biellls** [comment]
I will try to find time to finish it next weekend. If I don't you can close it.
Also I'm not sure how to use the travis ssh server for tests and sshook tests don't help a lot since they are very simple. I can do the tests with mocks if that is ok too.

**Comment by r39132** [comment]
@biellls Closing for now. Please reopen when you have time!

**Comment by iamdanielsun** [comment]
Sorry to resurrect an old PR, but are there any further developments here? I am receiving the following exception when trying to use SFTPHook: paramiko.ssh_exception.SSHException: Error reading SSH protocol banner

The package versions I have tested with:
pysftp: 0.2.8 (downgraded from 0.2.9 because of the issue outlined here: https://github.com/Yenthe666/auto_backup/issues/47 )
paramiko: 2.1.1 and 1.15.1

Edit: Alright I feel really stupid. The issue was that I had forgotten to change the port from 21 (FTP) to SFTP (22) in the Airflow web console. Everything works now. Thanks biellls!

**Pull request closed without merging**